packages/toolpad-app/src/components/DialogForm.tsx

```diff
@@ -1,7 +1,39 @@
+import * as React from 'react';
 import { styled } from '@mui/material';
 
-export default styled('form')({
+const StyledDialogForm = styled('form')({
   overflowY: 'auto',
   display: 'flex',
   flexDirection: 'column',
 });
+
+export interface DialogFormProps extends React.FormHTMLAttributes<HTMLFormElement> {
+  children: React.ReactNode;
+  onKeyDown?: (event: React.KeyboardEvent) => void;
+}
+
+export default ({ children, onKeyDown = () => {}, ...rest }: DialogFormProps) => {
+  const handleKeyDown = React.useCallback(
+    (event: React.KeyboardEvent) => {
+      const modifierKey = (event.metaKey || event.ctrlKey) && !event.shiftKey;
+
+      if (event.key === 'Enter' && modifierKey) {
+        const submitEvent = new Event('submit', {
+          cancelable: true,
+          bubbles: true,
+        });
+
+        event.currentTarget.dispatchEvent(submitEvent);
```

Comment:
Great idea. One drawback I see is that it will submit the form, even when the submit button is disabled (I noticed you have to add several guards across the form to account for this). Would it be possible to instead of dispatching an event on the form, query the DOM for a submit button and fire a click event? e.g.

```tsx
ref.current.querySelector('[type=submit]').click()
```

Not 100% sure this would work.

Reply:
I don't believe this is a right approach, IMO submit logic shouldn't be guarded by the submit button, instead we should always guard logic with the right condition inside the handler. While the above should work, I don't think that it should be implemented for the reason mentioned above. It is also possible to have forms without submit button entirely which could be "remotely" submitted, in that case it wouldn't work at all

Reply:
> I don't believe this is a right approach, IMO submit logic shouldn't be guarded by the submit button, instead we should always guard logic with the right condition inside the handler.

From a UX point of view, the Cmd+Enter is a shortcut for the submit button. So it shouldn't submit the form when the button is disabled. Emulating a click on the submit button is the closest I can think of to bening a shortcut for that button.
If you feel strongly about guarding the logic inside the handlers, then at least use an `invariant`. The handler should never be called when the submit button is disabled, so when it is called I want to see an error because that signifies a there has been a programming mistake. And we want these to surface, not being ignored.

> It is also possible to have forms without submit button entirely which could be "remotely" submitted, in that case it wouldn't work at all

I think we can safely say that for a good UX, all dialogs that have a form will have a submit button. 

I don't think I'm familiar with "remotely submitting forms", can you point me to some documentation about this feature?

Reply:
> From a UX point of view, the Cmd+Enter is a shortcut for the submit button

Thanks, you are right. I checked how the form behaves with a regular `<input />` when enter is pressed and it does properly regard submit button state. I've changed implementation as per your suggestion and it seems to work as expected now 👍 

> If you feel strongly about guarding the logic inside the handlers, then at least use an invariant

I've changed to usage of `invariant`. It's not that I feel strongly about current behaviour, I just believe that validation should happen inside the handler itself.

> I think we can safely say that for a good UX, all dialogs that have a form will have a submit button.

yeh, that's absolutely correct

> I don't think I'm familiar with "remotely submitting forms", can you point me to some documentation about this feature?

As per previous comment, this won't probably be applicable to our `Dialogs`, but an example to illustrate what I meant could be something like this - you have a form fields rendered in the main content area, but submit button is "taken" out of form and placed in a navigation bar or some other dom element which doesn't have access to trigger submit event. Upon pressing submit you would programatically trigger submit on the form, kicking in whatever the submit logic is defined. I don't know actually if "remote form submit" is correct term though, it might be misleading.

---

While we are on this subject I'd just like to share a couple of resources which explain why using disabled form buttons in general isn't good idea: 
- [one](https://www.smashingmagazine.com/2021/08/frustrating-design-patterns-disabled-buttons/)
- [two](https://medium.com/@tu.bui/dont-use-disabled-buttons-7541d0586ca1)
- [three](https://uxdesign.cc/why-heuristics-are-only-rules-of-thumb-the-case-of-the-disabled-button-4824958627e9) 
- [four](https://uxplanet.org/disabled-buttons-in-user-interface-4dafda3e6fe7)

tl;dr; - it is considered better to allow users to make mistake and show error explaining what's wrong, rather than preventing them to make the mistake without understanding why they can't proceed

IMO in our case a good example could be a component creation form - when I remove the name there is nothing in UI indicating why I can't proceed. So unless I can implicitly understand that the name is required, I'd be stuck
<img width="438" alt="image" src="https://user-images.githubusercontent.com/437214/197511626-97530c80-34e6-46da-8df5-27bc56639975.png">

Anyways, this is probably a topic for a different occasion

Reply:
> Anyways, this is probably a topic for a different occasion

I agree, it's a debate we should probably do outside of this ticket. A few notes:

1. Incidentally, I just fixed those node creation dialogs this morning, they now give proper feedback:
    
    <img width="485" alt="Screenshot 2022-10-24 at 13 13 46" src="https://user-images.githubusercontent.com/2109932/197514161-385d4273-3174-431c-be60-d821907164e0.png">

2. These blog posts touch on interesting points, but I don't think they're proposing the right UX neither. First of all, they present a false dichotomy, either have a disabled button and no validation on the inputs, or have a disabled button with validation on the inputs. But you can have validation as well as a disabled submit button. Second, my expectation when clicking a submit button is that "my work is done". It's very frustrating to click submit only to find I'm going to have to spend more time writing inputs. It's not that I don't want to spend the time, it's that I expect to be done when I click submit and can go on with the day, the button doesn't do what I expect it to do. For inputs that are offscreen, it's perfectly possible to accompany the submit button with a message, with a link that scrolls to the input.

Reply:
> But you can have validation as well as a disabled submit button

You can, but unless you interact with a field - validation message won't appear where with enabled submit button you provide a way to continue and "fail".

Anyways, a topic for another day..

Reply:
From a UX perspective, I would 👍 on **a default** of making the submit button clickable, even when the form is invalid to be able to raise a clear error message.

However, I think that it also makes sense to disable the button if there is a strong use case for it. For example with very simple forms. This GitHub comment texterea I'm commenting on and submitting with <kbd>Ctrl</kbd>+<kbd>Enter</kbd>:

<img width="293" alt="Screenshot 2022-10-25 at 19 12 53" src="https://user-images.githubusercontent.com/3165635/197839047-29c20455-8446-4986-b55a-008985febbb2.png">